### PR TITLE
Enable pickling for tensor and tensor based geometry

### PR DESCRIPTION
**Pull request by yuecideng**
This PR makes `Tensor`, `TensorMap` and tensor based geometry pickling.

Current behaviours:
- Tensor will be on the same device after de-serialization.
- Non contiguous tensor will be converted to contiguous tensor after de-serilization.

- - -
This change is [Reviewable](https://reviewable.io/reviews/isl-org/open3d/5509)

### CODE REVIEW

**Comment by yxlao** [comment]
LGTM after adding python unit tests.

**Comment by yuecideng** [comment]
- Add python unit tests for pickling support.
- Minor change `PointCloud::ToString()` output to be consistent with `TriangleMesh`:
```
PointCloud on CPU:0 [2 points (Float32)].
Attributes: normals (dtype = Float32, shape = {2, 3}).
```
@yxlao

File: .gitignore [hunk 1]
```diff
@@ -90,3 +90,6 @@ docs/Doxyfile
 docs/getting_started.rst
 docs/docker.rst
 docs/tensorboard.md
+
+# test
+*.pkl
```
Comment (yxlao) [inline .gitignore:95]: remove this

File: cpp/pybind/core/tensor.cpp [hunk 2]
```diff
@@ -356,6 +356,34 @@ void pybind_core_tensor(py::module& m) {
...
+                return py::make_tuple(t.GetDevice(),
+                                      TensorToPyArray(t.To(Device("CPU:0"))));
+            },
+            [](py::tuple t) {
+                // __setstate__
+                if (t.size() != 2) {
+                    utility::LogError(
+                            "Invalid state! Expecting a tuple of size 2.");
+                }
+                const Device& device = t[0].cast<Device>();
```
Comment (yxlao) [inline cpp/pybind/core/tensor.cpp:375]: Use `Device::IsAvailable()`. For example, when pickling, we might store `CUDA:1`, but another computer may only have `CUDA:0`. Same for others.

File: cpp/pybind/t/geometry/image.cpp [hunk 3]
```diff
@@ -101,6 +101,22 @@ void pybind_image(py::module &m) {
                  "tensor"_a);
     docstring::ClassMethodDocInject(m, "Image", "__init__",
                                     map_shared_argument_docstrings);
+
+    // Pickle support.
+    image.def(py::pickle(
+            [](const Image &image) {
+                // __getstate__
+                return py::make_tuple(image.AsTensor());
```
Comment (yxlao) [inline cpp/pybind/t/geometry/image.cpp:109]: Do we need to transfer to CPU?

File: cpp/pybind/t/geometry/image.cpp [hunk 4]
```diff
@@ -282,6 +298,26 @@ void pybind_image(py::module &m) {
...
+
+            // Pickling support.
+            .def(py::pickle(
+                    [](const RGBDImage &rgbd) {
+                        // __getstate__
+                        return py::make_tuple(rgbd.color_, rgbd.depth_,
+                                              rgbd.aligned_);
+                    },
+                    [](py::tuple t) {
+                        // __setstate__
```
Comment (yxlao) [inline cpp/pybind/t/geometry/image.cpp:310]: Same, do we need to check device as well?

File: cpp/pybind/core/device.cpp [hunk 5]
```diff
@@ -46,7 +46,20 @@ void pybind_core_device(py::module &m) {
...
-            .def("get_id", &Device::GetID);
+            .def("get_id", &Device::GetID)
+            .def(py::pickle(
+                    [](const Device &d) {
+                        return py::make_tuple(d.GetType(), d.GetID());
+                    },
+                    [](py::tuple t) {
+                        if (t.size() != 2) {
+                            utility::LogError(
+                                    "Invalid state! Expecting a tuple of size "
```
Comment (yxlao) [inline cpp/pybind/core/device.cpp:57]: Add a comment: "Cannot unpickle Device." It is useful for the user to know which class went wrong. Same for others.

File: cpp/pybind/core/tensor.cpp [hunk 6]
```diff
@@ -356,6 +356,34 @@ void pybind_core_tensor(py::module& m) {
...
+                if (device.IsCUDA() && !core::cuda::IsAvailable()) {
+                    utility::LogWarning(
+                            "CUDA is not available, tensor will be "
```
Comment (yxlao) [inline cpp/pybind/core/tensor.cpp:378]: Change this comment as well. Same for others.

File: python/test/t/geometry/test_image.py [hunk 7]
```diff
@@ -76,3 +77,12 @@ def test_buffer_protocol_cpu(device):
         im = im.to(device=device)
         dst_t = np.asarray(im.cpu())
         np.testing.assert_array_equal(src_t, dst_t)
+
+
+@pytest.mark.parametrize("device", list_devices())
+def test_pickle(device):
+    img = o3d.t.geometry.Image(o3c.Tensor.ones((10, 10, 3), o3c.uint8, device))
+    pickle.dump(img, open("img.pkl", "wb"))
```
Comment (yxlao) [inline python/test/t/geometry/test_image.py:85]: Use 
```python
import tempfile

with tempfile.TemporaryDirectory() as temp_dir:
    pickle_path = xxx
```

Same for others.

**Review by yxlao — CHANGES_REQUESTED** [review]

File: cpp/pybind/t/geometry/image.cpp (hunk 3 above)
Comment (yuecideng) [reply cpp/pybind/t/geometry/image.cpp:109]: The conversion to CPU is performed in `Tensor` side, so we don't need here.

File: cpp/pybind/t/geometry/image.cpp (hunk 4 above)
Comment (yuecideng) [reply cpp/pybind/t/geometry/image.cpp:310]: Same, the device will be checked by the tensor inside Image.

File: .gitignore (hunk 1 above)
Comment (yuecideng) [reply .gitignore]: Done.

File: cpp/pybind/core/tensor.cpp (hunk 2 above)
Comment (yuecideng) [reply cpp/pybind/core/tensor.cpp:376]: Done.

File: cpp/pybind/core/device.cpp (hunk 5 above)
Comment (yuecideng) [reply cpp/pybind/core/device.cpp]: Done.

File: python/test/t/geometry/test_image.py (hunk 7 above)
Comment (yuecideng) [reply python/test/t/geometry/test_image.py]: Done.

**Review by yxlao — APPROVED** [review]

**Comment by reyanshsolis** [comment]
Great work. It would be very helpful to add a brief documentation example on the Tensor page.
TODO: Check newly added features that are missing documentation and make a priority based checklist for the same.

**Comment by yuecideng** [comment]
> It would be very helpful to add a brief documentation example on the Tensor page

Sounds good, I will add it later.

**Pull request merged**